mlir/lib/Analysis/SliceAnalysis.cpp

```diff
@@ -91,7 +92,13 @@ static void getBackwardSliceImpl(Operation *op,
   if (options.filter && !options.filter(op))
     return;
 
-  for (const auto &en : llvm::enumerate(op->getOperands())) {
+  auto operands = op->getOperands();
+  SetVector<Value> valuesToFollow(operands.begin(), operands.end());
+  if (!options.omitUsesFromAbove) {
+    getUsedValuesDefinedAbove(op->getRegions(), valuesToFollow);
+  }
+
+  for (const auto &en : llvm::enumerate(valuesToFollow)) {
```

Comment:
The way you can make it "free" when there are no regions is to define this code as a lambda that operates on an `ValueRange`.

```
auto processValues = [&] (ValueRange valuesToFollow) {
  ...
};
```

And then dispatch like this:
```
processValues(op->getOperands());

   if (!options.omitUsesFromAbove) {
     SetVector<Value> valuesToFollow;
     getUsedValuesDefinedAbove(op->getRegions(), valuesToFollow);
     processValues(valuesToFollow);
   }
```

Reply:
Thank you, that's a good solution. 

I tweaked it a bit to take a `Value` so it can get passed to `visitValuesDefinedAbove` which takes a lambda with an `OpOperand*` arg. This prevents storing the values in a set (`backwardSlice->count` ensures they are still only visited once)